tests/integration/test_gpio_expander_cache.py

```
@@ -0,0 +1,122 @@
+        (digital_read_hw_pattern, 3),
+        (digital_read_cache_pattern, 3),
+        (digital_read_hw_pattern, 3),
+        (digital_read_cache_pattern, 3),
+        (digital_read_cache_pattern, 4),
+        (digital_read_hw_pattern, 3),
+        (digital_read_cache_pattern, 3),
+        (digital_read_hw_pattern, 10),
+        (digital_read_cache_pattern, 10),
+        # full cache reset here for testing
+        (digital_read_hw_pattern, 15),
+        (digital_read_cache_pattern, 15),
+        (digital_read_cache_pattern, 14),
+        (digital_read_hw_pattern, 14),
+        (digital_read_cache_pattern, 14),
+    ]
+    # Flatten the log order for easier processing
+    log_order: list[tuple[re.Pattern, int]] = [
+        item
+        for sublist in log_order
+        for item in (sublist if isinstance(sublist, list) else [sublist])
+    ]
+
+    index = 0
+
+    def check_output(line: str) -> None:
+        """Check log output for expected messages."""
+        nonlocal index
+        if logs_done.is_set():
+            return
+
+        clean_line = re.sub(r"\x1b\[[0-9;]*m", "", line)
+
+        if "digital_read" in clean_line:
+            if index < len(log_order):
+                pattern, expected_pin = log_order[index]
+                match = pattern.search(line)
+                if match:
+                    pin = int(match.group(1))
+                    if pin == expected_pin:
+                        index += 1
+                    else:
+                        print(
+                            f"Unexpected pin number. Expected {expected_pin}, got {pin}"
+                        )
+                        logs_done.set()
+                else:
+                    print(f"Log line did not match next expected pattern: {clean_line}")
+                    logs_done.set()
+            else:
+                print(f"Received unexpected log line: {clean_line}")
+                index += 1
+                logs_done.set()
```

```suggestion
        if "digital_read" in clean_line:
            if index >= len(log_order):
                print(f"Received unexpected log line: {clean_line}")
                logs_done.set()
                return
            
            pattern, expected_pin = log_order[index]
            match = pattern.search(clean_line)
            
            if not match:
                print(f"Log line did not match next expected pattern: {clean_line}")
                logs_done.set()
                return
            
            pin = int(match.group(1))
            if pin != expected_pin:
                print(f"Unexpected pin number. Expected {expected_pin}, got {pin}")
                logs_done.set()
                return
            
            index += 1
```

---

probably pytest.fail instead of print though?

---

Wasnt sure about failing inside the callback function, but will give it a shot